File: game_recorder.js
/**
 * Flutter Stock Exchange Game - ML Training Data Recorder (JavaScript)
 * Browser-side recorder that creates standardized notation for ML analysis
 */

class FlutterGameRecorder {
    constructor() {
        this.currentGameId = null;
        this.turnCounter = 0;
        this.moves = [];
        this.gameStates = [];
        this.recordingEnabled = true;
    }

    // Standardized notation creation
    createNotation(action, player, target = "", value = 0, extra = "") {
        const parts = [action, player];
        if (target) parts.push(target);
        if (value) parts.push(value.toString());
        if (extra) parts.push(extra);
        return parts.join(":");
    }

    // Start recording a new game
    startGame(winTarget, playerCount, personalities) {
        this.currentGameId = `game_${Date.now()}`;
        this.turnCounter = 0;
        this.moves = [];
        this.gameStates = [];
        
        const gameStart = {
            gameId: this.currentGameId,
            timestamp: new Date().toISOString(),
            winTarget: winTarget,
            playerCount: playerCount,
            personalities: personalities
        };
        
        console.log(`🎮 ML Recording started: ${this.currentGameId}`);
        console.log(`📊 Players: ${playerCount}, Target: £${winTarget}`);
        
        // Store game start in localStorage for now
        localStorage.setItem(`game_${this.currentGameId}`, JSON.stringify(gameStart));
        
        return this.currentGameId;
    }

    // Record dice roll
    recordDiceRoll(diceValue, playerColor, newPosition) {
        if (!this.recordingEnabled) return;
        
        this.turnCounter++;
        const notation = this.createNotation("R", playerColor, "move", diceValue, newPosition);
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "R",
            player: playerColor,
            target: "move",
            value: diceValue,
            context: { newPosition: newPosition }
        });
        
        console.log(`🎲 ML: ${notation}`);
    }

    // Record buy transaction
    recordBuy(buyerColor, targetCompany, price, sharesCount = 1) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("B", buyerColor, targetCompany, price, sharesCount);
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "B",
            player: buyerColor,
            target: targetCompany,
            value: price,
            context: { shares: sharesCount, totalCost: price + 5 }
        });
        
        console.log(`💰 ML: ${notation}`);
    }

    // Record sell transaction
    recordSell(sellerColor, targetCompany, price, sharesCount = 1) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("S", sellerColor, targetCompany, price, sharesCount);
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "S",
            player: sellerColor,
            target: targetCompany,
            value: price,
            context: { shares: sharesCount }
        });
        
        console.log(`💸 ML: ${notation}`);
    }

    // Record dividend
    recordDividend(playerColor, dividendType, amount) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("D", playerColor, dividendType, amount);
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "D",
            player: playerColor,
            target: dividendType,
            value: amount,
            context: { dividendType: dividendType }
        });
        
        console.log(`💵 ML: ${notation}`);
    }

    // Record market event
    recordMarketEvent(playerColor, eventType, effect, value = 0) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("M", playerColor, eventType, value, effect);
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "M",
            player: playerColor,
            target: eventType,
            value: value,
            context: { effect: effect }
        });
        
        console.log(`📰 ML: ${notation}`);
    }

    // Record round end
    recordRoundEnd(winnerColor, winningPosition) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("E", winnerColor, "round", winningPosition, "win");
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "E",
            player: winnerColor,
            target: "round",
            value: winningPosition,
            context: { roundWin: true }
        });
        
        console.log(`🏁 ML: ${notation}`);
    }

    // Record game win
    recordGameWin(winnerColor, finalWealth) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("W", winnerColor, "game", finalWealth);
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "W",
            player: winnerColor,
            target: "game",
            value: finalWealth,
            context: { gameWin: true }
        });
        
        console.log(`🏆 ML: ${notation}`);
    }

    // Record complete game state for ML feature engineering
    recordGameState(gameState, currentPlayer) {
        if (!this.recordingEnabled) return;
        
        const state = {
            turn: this.turnCounter,
            timestamp: new Date().toISOString(),
            currentPlayer: currentPlayer,
            playerCash: gameState.playerCash,
            playerShares: gameState.playerShares,
            playerPositions: gameState.playerPositions,
            travelerPositions: gameState.travelerPositions,
            personalities: gameState.playerPersonalities
        };
        
        this.gameStates.push(state);
        
        // Compact notation for game state
        const cashTotal = Object.values(gameState.playerCash).reduce((a, b) => a + b, 0);
        const notation = this.createNotation("GS", currentPlayer, "state", cashTotal, this.turnCounter);
        
        console.log(`📊 ML State: ${notation}`);
    }

    // Record AI decision reasoning
    recordAIDecision(playerColor, decision, reasoning, personality = null) {
        if (!this.recordingEnabled) return;
        
        const notation = this.createNotation("AI", playerColor, decision.action, 
                                           decision.company || 0, reasoning.substring(0, 20));
        
        this.recordMove({
            turn: this.turnCounter,
            notation: notation,
            action: "AI",
            player: playerColor,
            target: decision.action,
            value: 0,
            context: { 
                decision: decision,
                reasoning: reasoning,
                personality: personality
            }
        });
        
        console.log(`🤖 ML: ${notation}`);
    }

    // Internal method to record moves
    recordMove(moveData) {
        try {
            // Safely clone gameState if present
            if (moveData.gameState) {
                moveData.gameState = this.safeCloneGameState(moveData.gameState);
            }
            
            moveData.gameId = this.currentGameId;
            moveData.timestamp = new Date().toISOString();
            this.moves.push(moveData);
        } catch (error) {
            console.warn('ML Recording error:', error.message);
        }
    }

    // Safely clone game state to avoid circular reference errors
    safeCloneGameState(gameState) {
        try {
            return JSON.parse(JSON.stringify(gameState));
        } catch (error) {
            // If JSON fails, create a safe subset
            return {
                playerPositions: gameState.playerPositions || {},
                travelerPositions: gameState.travelerPositions || {},
                playerCash: gameState.playerCash || {},
                playerShares: gameState.playerShares || {},
                playerPersonalities: gameState.playerPersonalities || {},
                activeColors: gameState.activeColors || [],
                currentPlayer: gameState.currentPlayer || 0,
                roundInProgress: gameState.roundInProgress || false
            };
        }
    }

    // Export data for ML analysis
    exportGameData() {
        const data = {
            gameId: this.currentGameId,
            moves: this.moves,
            gameStates: this.gameStates,
            moveCount: this.moves.length,
            totalTurns: this.turnCounter
        };
        
        // Save to localStorage
        localStorage.setItem(`moves_${this.currentGameId}`, JSON.stringify(data));
        
        console.log(`📤 ML Data exported: ${this.moves.length} moves, ${this.gameStates.length} states`);
        return data;
    }

    // Convert to standardized formats
    exportAsChessStyle() {
        // Create chess-like notation string
        return this.moves.map(move => move.notation).join(" ");
    }

    exportAsJSON() {
        return JSON.stringify(this.exportGameData(), null, 2);
    }

    exportAsCSV() {
        const headers = "turn,notation,action,player,target,value,timestamp";
        const rows = this.moves.map(m => 
            `${m.turn},${m.notation},${m.action},${m.player},${m.target},${m.value},${m.timestamp}`
        );
        return [headers, ...rows].join("\n");
    }

    // Download exported data
    downloadData(format = "json") {
        let data, filename, mimeType;
        
        switch(format) {
            case "json":
                data = this.exportAsJSON();
                filename = `flutter_game_${this.currentGameId}.json`;
                mimeType = "application/json";
                break;
            case "csv":
                data = this.exportAsCSV();
                filename = `flutter_game_${this.currentGameId}.csv`;
                mimeType = "text/csv";
                break;
            case "chess":
                data = this.exportAsChessStyle();
                filename = `flutter_game_${this.currentGameId}.txt`;
                mimeType = "text/plain";
                break;
        }
        
        const blob = new Blob([data], { type: mimeType });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = filename;
        a.click();
        URL.revokeObjectURL(url);
        
        console.log(`💾 Downloaded: ${filename}`);
    }

    // Print comprehensive statistics
    printStats() {
        const actions = this.moves.reduce((acc, move) => {
            acc[move.action] = (acc[move.action] || 0) + 1;
            return acc;
        }, {});
        
        console.log("=== COMPREHENSIVE GAME STATISTICS ===");
        console.log(`🎮 Game ID: ${this.currentGameId}`);
        console.log(`⏰ Turns: ${this.turnCounter}`);
        console.log(`📊 Total Moves: ${this.moves.length}`);
        console.log(`🎯 Recording Status: ${this.recordingEnabled ? 'ACTIVE' : 'PAUSED'}`);
        
        console.log("\n📈 Action Breakdown:");
        Object.entries(actions).forEach(([action, count]) => {
            console.log(`  ${action}: ${count}`);
        });
        
        // Detailed analysis
        this.printDetailedAnalysis();
        
        console.log("===============================");
        return actions;
    }

    // Detailed analysis for ML insights
    printDetailedAnalysis() {
        const moves = this.moves;
        if (moves.length === 0) return;
        
        console.log("\n🔍 DETAILED ANALYSIS:");
        
        // Dice roll analysis
        const diceRolls = moves.filter(m => m.action === 'dice_roll');
        if (diceRolls.length > 0) {
            const avgRoll = diceRolls.reduce((sum, m) => sum + m.numberResult, 0) / diceRolls.length;
            console.log(`🎲 Dice Rolls: ${diceRolls.length} (avg: ${avgRoll.toFixed(2)})`);
        }
        
        // Trading analysis
        const trades = moves.filter(m => m.action.includes('shares'));
        const buys = moves.filter(m => m.action.includes('buy'));
        const sells = moves.filter(m => m.action.includes('sell'));
        console.log(`💰 Trading: ${trades.length} total (${buys.length} buys, ${sells.length} sells)`);
        
        // Special events
        const slumps = moves.filter(m => m.action === 'slump_effect');
        const marketNews = moves.filter(m => m.action === 'market_news_triggered');
        const dividends = moves.filter(m => m.action === 'dividend_paid');
        const bankruptcies = moves.filter(m => m.action === 'bankruptcy');
        
        console.log(`📰 Market Events: ${marketNews.length} market news, ${slumps.length} slumps`);
        console.log(`💵 Dividends: ${dividends.length} payments`);
        if (bankruptcies.length > 0) {
            console.log(`💀 Bankruptcies: ${bankruptcies.length}`);
        }
        
        // Insurance usage
        const antiSlump = moves.filter(m => m.action === 'anti_slump_used');
        if (antiSlump.length > 0) {
            console.log(`🛡️ Anti-Slump Cards Used: ${antiSlump.length}`);
        }
        
        // Round analysis
        const rounds = moves.filter(m => m.action === 'round_end');
        if (rounds.length > 0) {
            console.log(`🏁 Rounds Completed: ${rounds.length}`);
        }
    }

    // Generate comprehensive report
    generateReport() {
        const report = {
            gameId: this.currentGameId,
            totalMoves: this.moves.length,
            totalStates: this.gameStates.length,
            recordingDuration: this.moves.length > 0 ? 
                new Date(this.moves[this.moves.length - 1].timestamp) - new Date(this.moves[0].timestamp) : 0,
            
            // Action counts
            actionCounts: this.moves.reduce((acc, move) => {
                acc[move.action] = (acc[move.action] || 0) + 1;
                return acc;
            }, {}),
            
            // Game flow metrics
            diceRolls: this.moves.filter(m => m.action === 'dice_roll').length,
            tradesExecuted: this.moves.filter(m => m.action.includes('shares')).length,
            specialEvents: this.moves.filter(m => 
                ['slump_effect', 'market_news_triggered', 'dividend_paid'].includes(m.action)
            ).length,
            
            // Financial metrics
            totalDividendsPaid: this.moves
                .filter(m => m.action === 'dividend_paid')
                .reduce((sum, m) => sum + (m.totalDividendsPaid || 0), 0),
            
            // Risk events
            slumpEvents: this.moves.filter(m => m.action === 'slump_effect').length,
            antiSlumpUsage: this.moves.filter(m => m.action === 'anti_slump_used').length,
            bankruptcies: this.moves.filter(m => m.action === 'bankruptcy').length,
            
            // Game completion
            roundsCompleted: this.moves.filter(m => m.action === 'round_end').length,
            gameEnded: this.moves.some(m => m.action === 'game_end'),
            
            // Timestamps
            startTime: this.moves.length > 0 ? this.moves[0].timestamp : null,
            endTime: this.moves.length > 0 ? this.moves[this.moves.length - 1].timestamp : null
        };
        
        return report;
    }
}

// Add missing methods needed by the HTML interface
FlutterGameRecorder.prototype.startNewGame = function() {
    this.currentGameId = `game_${Date.now()}`;
    this.turnCounter = 0;
    this.moves = [];
    this.gameStates = [];
    console.log(`🎮 ML Recording started: ${this.currentGameId}`);
    return this.currentGameId;
};

FlutterGameRecorder.prototype.endGame = function(winnerId, finalScore) {
    this.recordGameWin(winnerId, finalScore);
    console.log(`🏆 Game ended: ${winnerId} wins with £${finalScore}`);
};

FlutterGameRecorder.prototype.exportData = function() {
    return this.exportAsJSON();
};

FlutterGameRecorder.prototype.toggleRecording = function() {
    this.recordingEnabled = !this.recordingEnabled;
    console.log(`🎬 Recording ${this.recordingEnabled ? 'enabled' : 'disabled'}`);
    return this.recordingEnabled;
};

// Global recorder instance
window.mlRecorder = new FlutterGameRecorder(); 

File: game_recorder.py
#!/usr/bin/env python3
"""
Flutter Stock Exchange Game - ML Training Data Recorder
Creates standardized notation system for machine learning analysis
"""

import sqlite3
import json
import datetime
from pathlib import Path
from typing import Dict, List, Optional

class FlutterGameRecorder:
    """Records Flutter game moves in standardized notation for ML training"""
    
    def __init__(self, db_path: str = "game.db"):
        self.db_path = db_path
        self.init_database()
    
    def init_database(self):
        """Initialize SQLite database with game recording schema"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Games table
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS games (
                game_id INTEGER PRIMARY KEY AUTOINCREMENT,
                start_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                end_time TIMESTAMP,
                win_target INTEGER,
                player_count INTEGER,
                winner TEXT,
                total_turns INTEGER,
                game_duration_minutes REAL,
                personalities TEXT  -- JSON of player:personality mapping
            )
        """)
        
        # Moves table - standardized notation
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS moves (
                move_id INTEGER PRIMARY KEY AUTOINCREMENT,
                game_id INTEGER,
                turn_number INTEGER,
                move_notation TEXT,  -- Standardized notation like "R:4:yellow:130"
                timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                player_color TEXT,
                action_type TEXT,    -- R, B, S, D, M, E, W
                target TEXT,         -- Company or position
                value INTEGER,       -- Price, amount, dice roll
                context TEXT,        -- Additional data as JSON
                FOREIGN KEY (game_id) REFERENCES games (game_id)
            )
        """)
        
        # Game states table - for ML feature engineering
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS game_states (
                state_id INTEGER PRIMARY KEY AUTOINCREMENT,
                game_id INTEGER,
                turn_number INTEGER,
                player_color TEXT,
                cash INTEGER,
                share_values TEXT,     -- JSON of company:shares owned
                company_positions TEXT, -- JSON of company:price positions
                traveler_positions TEXT, -- JSON of company:traveler positions
                personality TEXT,
                decision_reasoning TEXT,
                FOREIGN KEY (game_id) REFERENCES games (game_id)
            )
        """)
        
        # ML Features table - derived features for training
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS ml_features (
                feature_id INTEGER PRIMARY KEY AUTOINCREMENT,
                game_id INTEGER,
                player_color TEXT,
                turn_number INTEGER,
                -- Financial features
                net_worth INTEGER,
                cash_ratio REAL,
                portfolio_diversity REAL,
                -- Market features
                market_volatility REAL,
                position_momentum REAL,
                -- Personality features
                risk_tolerance REAL,
                aggressiveness REAL,
                trading_frequency REAL,
                -- Decision features
                action_taken TEXT,
                action_value INTEGER,
                success_metric REAL,  -- Calculated post-action
                FOREIGN KEY (game_id) REFERENCES games (game_id)
            )
        """)
        
        conn.commit()
        conn.close()
        print(f"✅ Database initialized: {self.db_path}")
    
    def start_game(self, win_target: int, player_count: int, personalities: Dict[str, str]) -> int:
        """Start recording a new game"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            INSERT INTO games (win_target, player_count, personalities)
            VALUES (?, ?, ?)
        """, (win_target, player_count, json.dumps(personalities)))
        
        game_id = cursor.lastrowid
        conn.commit()
        conn.close()
        
        print(f"🎮 Started recording game {game_id}")
        return game_id
    
    def record_move(self, game_id: int, turn: int, notation: str, player: str, 
                   action_type: str, target: str = "", value: int = 0, context: Dict = None):
        """Record a single move in standardized notation"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            INSERT INTO moves (game_id, turn_number, move_notation, player_color, 
                             action_type, target, value, context)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        """, (game_id, turn, notation, player, action_type, target, value, 
              json.dumps(context) if context else None))
        
        conn.commit()
        conn.close()
    
    def record_game_state(self, game_id: int, turn: int, player: str, 
                         cash: int, shares: Dict, positions: Dict, travelers: Dict,
                         personality: str, reasoning: str = ""):
        """Record complete game state for ML feature engineering"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        cursor.execute("""
            INSERT INTO game_states (game_id, turn_number, player_color, cash,
                                   share_values, company_positions, traveler_positions,
                                   personality, decision_reasoning)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        """, (game_id, turn, player, cash, json.dumps(shares), 
              json.dumps(positions), json.dumps(travelers), personality, reasoning))
        
        conn.commit()
        conn.close()
    
    def end_game(self, game_id: int, winner: str, total_turns: int):
        """Complete game recording"""
        conn = sqlite3.connect(self.db_path)
        cursor = conn.cursor()
        
        # Calculate game duration
        cursor.execute("SELECT start_time FROM games WHERE game_id = ?", (game_id,))
        start_time = cursor.fetchone()[0]
        duration = (datetime.datetime.now() - datetime.datetime.fromisoformat(start_time)).total_seconds() / 60
        
        cursor.execute("""
            UPDATE games 
            SET end_time = CURRENT_TIMESTAMP, winner = ?, total_turns = ?, game_duration_minutes = ?
            WHERE game_id = ?
        """, (winner, total_turns, duration, game_id))
        
        conn.commit()
        conn.close()
        
        print(f"🏆 Game {game_id} completed - Winner: {winner} in {total_turns} turns")

# Notation helper functions
def create_notation(action: str, player: str, target: str = "", value: int = 0, extra: str = "") -> str:
    """Create standardized notation string"""
    parts = [action, player]
    if target: parts.append(target)
    if value: parts.append(str(value))
    if extra: parts.append(extra)
    return ":".join(parts)

def parse_notation(notation: str) -> Dict:
    """Parse notation string back into components"""
    parts = notation.split(":")
    return {
        "action": parts[0] if len(parts) > 0 else "",
        "player": parts[1] if len(parts) > 1 else "",
        "target": parts[2] if len(parts) > 2 else "",
        "value": int(parts[3]) if len(parts) > 3 and parts[3].isdigit() else 0,
        "extra": parts[4] if len(parts) > 4 else ""
    }

# Example notation patterns:
"""
DICE ROLLS:
R:4:yellow:130        # Roll 4, yellow player, move to position 130

TRADING:
B:red:yellow:120      # Buy red shares, by yellow, at price 120
S:blue:green:150      # Sell blue shares, by green, at price 150

DIVIDENDS:
D:20%:yellow:40       # 20% dividend, yellow receives £40
D:10%:red:25          # 10% dividend, red receives £25

MARKET EVENTS:
M:news:yellow:advance:6    # Market news, yellow advances 6 spaces
M:slump:red:fall:60        # Slump event, red falls 60 points

ROUND EVENTS:
E:yellow:220:win      # Round end, yellow reached £220, won round
W:green:850           # Win condition, green has £850 total wealth

GAME STATES (for ML):
GS:yellow:300:red:2,blue:1:pos:red:120,blue:110  # Game state snapshot
"""

if __name__ == "__main__":
    # Test the recorder
    recorder = FlutterGameRecorder()
    
    # Example usage
    game_id = recorder.start_game(600, 4, {
        "yellow": "conservativeInvestor", 
        "red": "aggressiveSpeculator",
        "blue": "balancedTrader",
        "green": "dividendHunter"
    })
    
    # Example moves
    recorder.record_move(game_id, 1, "R:4:yellow:130", "yellow", "R", "move", 4)
    recorder.record_move(game_id, 1, "B:red:yellow:120", "yellow", "B", "red", 120)
    
    print("✅ Test recording completed") 